File: AndroidStudio/ChallengeTogether/app/src/main/java/com/yjy/challengetogether/alarm/MyFirebaseMessagingService.java
package com.yjy.challengetogether.alarm;

import android.text.TextUtils;

import androidx.annotation.NonNull;

import com.google.firebase.messaging.FirebaseMessagingService;
import com.google.firebase.messaging.RemoteMessage;
import com.yjy.challengetogether.util.Util;

public class MyFirebaseMessagingService extends FirebaseMessagingService {

    Util util = new Util(MyFirebaseMessagingService.this);

    @Override
    public void onNewToken(@NonNull String token) {
        super.onNewToken(token);

        // 앱을 디바이스에서 처음 실행하거나 토큰 값이 변경될때 실행된다.
        // 하지만 아직 로그인도 하지 않은 사용자의 토큰값을 받아와도 DB에 저장할 수 없으니 HomeFragment에서 DB에 저장하도록 설정

        // 아래는 만약 사용자가 접속하지 않는 사이 토큰이 만료되어 재발급 됐을 경우에 유저가 접속하기 전 알람을 수신하게 하기 위한 재발급 및 서버 동기화 코드
        if (!TextUtils.isEmpty(util.getSessionKey())) {
            util.registerTokenToServer(token);
        }
    }

    @Override
    public void onMessageReceived(@NonNull RemoteMessage remoteMessage) {
        super.onMessageReceived(remoteMessage);
        // 메시지 데이터에서 제목과 내용을 추출

        String type = remoteMessage.getData().get("type");
        String title = remoteMessage.getData().get("title");
        String message = remoteMessage.getData().get("body");

        // https://firebase.google.com/docs/cloud-messaging/concept-options?hl=ko 참조
        // json 형식에 따라 커스텀 해서 받아올 수도 있고 기본 제공 json으로 받아올 수도 있는듯
        // String message = remoteMessage.getData().get("");

        // 지정된 알림 설정이 꺼져있으면 알림을 울리지 않음
        if (!util.getPushSettings(type)) {
            return;
        }

        NotificationHelper notificationHelper = new NotificationHelper(MyFirebaseMessagingService.this);
        notificationHelper.createNotification(type, title, message, remoteMessage.getData().get("idx"));
    }
}

File: app/src/main/java/com/yjy/challengetogether/fcm/MyFirebaseMessagingService.kt
package com.yjy.challengetogether.fcm

import com.google.firebase.messaging.FirebaseMessagingService
import com.google.firebase.messaging.RemoteMessage
import com.yjy.challengetogether.di.FcmEntryPoint
import com.yjy.data.auth.api.AppLockRepository
import com.yjy.data.user.api.UserRepository
import com.yjy.model.common.notification.Notification
import com.yjy.model.common.notification.NotificationSettingFlags
import com.yjy.model.common.notification.NotificationType
import com.yjy.platform.notifications.NotificationHelper
import dagger.hilt.EntryPoints
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import timber.log.Timber
import java.time.LocalDateTime

class MyFirebaseMessagingService : FirebaseMessagingService() {

    private val serviceScope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    private val notificationMapper: NotificationMapper by lazy {
        EntryPoints.get(applicationContext, FcmEntryPoint::class.java).notificationMapper()
    }

    private val userRepository: UserRepository by lazy {
        EntryPoints.get(applicationContext, FcmEntryPoint::class.java).userRepository()
    }

    private val appLockRepository: AppLockRepository by lazy {
        EntryPoints.get(applicationContext, FcmEntryPoint::class.java).appLockRepository()
    }

    override fun onNewToken(token: String) {
        Timber.d("Refreshed token: $token")
    }

    override fun onMessageReceived(remoteMessage: RemoteMessage) {
        super.onMessageReceived(remoteMessage)
        if (remoteMessage.data.isEmpty()) return

        val notification = extractNotification(remoteMessage.data) ?: return

        serviceScope.launch {
            handleFcmMessage(notification)
        }
        logNotification(notification)
    }

    private fun extractNotification(data: Map<String, String>): Notification? {
        val header = data[HEADER] ?: return null
        val body = data[BODY] ?: return null
        val type = data[TYPE] ?: return null
        val linkIdx = data[LINK_IDX]?.toIntOrNull() ?: return null

        return Notification(
            id = linkIdx,
            createdDateTime = LocalDateTime.now(),
            header = header,
            body = body,
            type = NotificationType.fromString(type),
            linkId = linkIdx,
        )
    }

    private suspend fun handleFcmMessage(notification: Notification) {
        if (isNotificationMuted(notification.type, notification.linkId)) return
        if (isNotificationSettingDisabled(notification.type)) return

        NotificationHelper.postNotification(
            context = this,
            notification = with(notificationMapper) {
                notification.toPlatformNotification(
                    shouldHideContent = appLockRepository.shouldHideNotificationContents.first(),
                )
            },
        )
    }

    private suspend fun isNotificationSettingDisabled(notificationType: NotificationType): Boolean {
        val settings = userRepository.getNotificationSettings()

        if (!NotificationSettingFlags.isEnabled(settings, NotificationSettingFlags.ALL)) {
            return true
        }

        val settingFlag = when (notificationType) {
            NotificationType.CHALLENGE_COMPLETE,
            -> NotificationSettingFlags.GOAL_ACHIEVED

            NotificationType.CHALLENGE_JOIN,
            NotificationType.CHALLENGE_LEAVE,
            -> NotificationSettingFlags.WAITING_ROOM_PARTICIPANTS

            NotificationType.CHALLENGE_START,
            NotificationType.CHALLENGE_DELETE,
            -> NotificationSettingFlags.WAITING_ROOM_UPDATE

            NotificationType.CHALLENGE_RESET,
            NotificationType.CHALLENGE_GIVE_UP,
            NotificationType.CHALLENGE_FORCE_REMOVE,
            -> NotificationSettingFlags.CHALLENGE_ROOM_PROGRESS

            NotificationType.CHALLENGE_STARTED_POST,
            NotificationType.CHALLENGE_WAITING_POST,
            -> NotificationSettingFlags.CHALLENGE_BOARD

            NotificationType.COMMUNITY_POST -> NotificationSettingFlags.COMMUNITY_COMMENT
            NotificationType.COMMUNITY_COMMENT -> NotificationSettingFlags.COMMUNITY_REPLY
            NotificationType.COMMON -> return false
        }

        return !NotificationSettingFlags.isEnabled(settings, settingFlag)
    }

    private suspend fun isNotificationMuted(
        notificationType: NotificationType,
        notificationLinkId: Int,
    ) = when (notificationType) {
        NotificationType.CHALLENGE_WAITING_POST,
        NotificationType.CHALLENGE_STARTED_POST,
        -> isChallengeBoardMuted(notificationLinkId)

        NotificationType.COMMUNITY_POST,
        NotificationType.COMMUNITY_COMMENT,
        -> isCommunityPostMuted(notificationLinkId)

        else -> false
    }

    private suspend fun isChallengeBoardMuted(challengeBoardId: Int): Boolean =
        userRepository.getMutedChallengeBoards().contains(challengeBoardId)

    private suspend fun isCommunityPostMuted(communityPostId: Int): Boolean =
        userRepository.getMutedCommunityPosts().contains(communityPostId)

    private fun logNotification(notification: Notification) {
        with(notification) {
            Timber.d("Received FCM Notification - ID: $id")
            Timber.d("Received FCM Notification - Header: $header")
            Timber.d("Received FCM Notification - Body: $body")
            Timber.d("Received FCM Notification - Type: ${type.type}")
            Timber.d("Received FCM Notification - LinkIndex: $linkId")
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        serviceScope.cancel()
    }

    private companion object {
        const val HEADER = "header"
        const val BODY = "body"
        const val TYPE = "type"
        const val LINK_IDX = "linkIdx"
    }
}
